snuba/query/snql/parser.py

```diff
@@ -144,57 +166,72 @@ class OrTuple(NamedTuple):
     name: str
+    sample_rate: Optional[float]
 
 
 class SnQLVisitor(NodeVisitor):
     """
     Builds Snuba AST expressions from the Parsimonious parse tree.
     """
 
-    def visit_query_exp(self, node: Node, visited_children: Iterable[Any]) -> Query:
-        match, where, collect, groupby, having, orderby, limit = visited_children
-        # check for empty clauses
-        if isinstance(groupby, Node):
-            groupby = None
-        if isinstance(orderby, Node):
-            orderby = None
-        if isinstance(having, Node):
-            having = None
-        if isinstance(limit, Node):
-            limit = None
-
-        selected_columns = collect
-        _groupby = None
-        if groupby:
-            selected_columns += groupby
-            _groupby = [g.expression for g in groupby]
-
-        return Query(
-            body={},
-            from_clause=match,
-            selected_columns=selected_columns,
-            array_join=None,
-            condition=where,
-            prewhere=None,
-            groupby=_groupby,
-            having=having,
-            order_by=orderby,
-            limit=limit,
-        )
+    def visit_query_exp(
+        self, node: Node, visited_children: Iterable[Any]
+    ) -> LogicalQuery:
+        args: MutableMapping[str, Any] = {
+            "array_join": None,
+        }
+        (
+            data_source,
+            args["selected_columns"],
+            args["groupby"],
+            args["condition"],
+            args["having"],
+            args["order_by"],
+            args["limitby"],
+            args["limit"],
+            args["offset"],
+            args["granularity"],
+            args["totals"],
+            _,
+        ) = visited_children
+
+        keys = list(args.keys())
+        for k in keys:
+            if isinstance(args[k], Node):
+                del args[k]
+
+        args.update({"body": {}, "prewhere": None, "from_clause": data_source})
+        if isinstance(data_source, QueryEntity):
+            # TODO: How sample rate gets stored needs to be addressed in a future PR
+            args["sample"] = data_source.sample
+
+        if "groupby" in args:
+            if "selected_columns" not in args:
+                args["selected_columns"] = []
+            args["selected_columns"] += args["groupby"]
+            args["groupby"] = map(lambda gb: gb.expression, args["groupby"])
+
+        return LogicalQuery(**args)
 
     def visit_match_clause(
-        self, node: Node, visited_children: Tuple[Any, Any, Any, EntityTuple, Any],
+        self, node: Node, visited_children: Tuple[Any, Any, Any, EntityTuple],
     ) -> QueryEntity:
-        _, _, _, match, _ = visited_children
+        _, _, _, match = visited_children
         key = EntityKey(match.name)
-        query_entity = QueryEntity(key, get_entity(key).get_data_model())
+        query_entity = QueryEntity(
+            key, get_entity(key).get_data_model(), match.sample_rate
+        )
```

Comment:
Question about queries without joins. Do we expect clients to assign aliases to to the entity even if there is no join in the query? If yes, are we fully qualifying the columns in  the query and is the response the user gets supposed to have column names prepended with the table alias or not ?

Reply:
All the language specifications show the entity in the MATCH clause with an alias. So the alias is required in the MATCH clause. However having that alias won't be required if there is no join. I think the user can choose whether they want the columns to have the entity alias or not. We should be respecting it either way.

Reply:
As long as we produce the same AST in both cases (the user providing aliases or not) it will be fine, otherwise it will be a mess for query processors. I think there are two points to address:
1) ensure consistency in the AST whether the user adds aliases or not
2) possibly have consistency in the AST between composite and simple query types. Composite queries must have them, so either we manage the two cases differently (simple vs composite) or we add aliases for simple queries as well, which means modifying the from clause structure since there is no way to store a table alias in a simple query.
2b) if we decide that the simple query won't have aliases, then we need to ensure all subqueries (that are not composite) will not have aliases.

Reply:
I'm going to make a note to add this in subsequent PRs. Right now I think it makes sense to add the alias to the from clause structure for simple queries. That avoids having two cases and it's better to have more data than less.

Reply:
I am ok in fully qualifying everything